### PR TITLE
fix(stepfunctions): map state validation fix

### LINKED ISSUE #4137: Support for Dynamic StepFunction Dynamic Batching with Map and Iterator
In regards to this blog post: https://aws.amazon.com/blogs/aws/new-step-functions-support-for-dynamic-parallelism/ - AWS StepFunctions now support Dynamic Parallelism (released publicly today AFAICT). Here is the relevant section in the state-languages spec: https://states-language.net/spec.html#map-state

Like [`Task`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-stepfunctions.Task.html), and [`Parallel`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-stepfunctions.Parallel.html), I believe this would be a new [`Construct`](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-stepfunctions-readme.html), with an [`IChainable`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-stepfunctions.IChainable.html) interface for `Map` and `Iterator`

### Use Case

I would like to use this in an in-development project that's already taking advantage of aws-cdk to create a workflow with stepfunctions. I'm currently iterating on payloads that can be executed in parallel, dynamically. I've steered away from trying to manage the state around a StateMachineFragment, but could go down that path until aws-cdk includes the aformentioned constructs.

### Proposed Solution

Implement `Map` and `Iterator` constructs. (More is likely needed).

### Other

supporting links:
- https://aws.amazon.com/blogs/aws/new-step-functions-support-for-dynamic-parallelism/ - AWS blog post announcing support
- https://states-language.net/spec.html#map-state - State Language spec : map-state

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by apalumbo**
Closes #4137, #4417

- Fixed validation for map state
- Also added validation for maxConcurremcy

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rhboyd** [comment]
Why is maxConcurrency mandatory? It should be optional with a default to like 10 or 100 if the user doesn’t supply it.

**Comment by apalumbo** [comment]
It is not mandatory, is number | undefined ,   in this way you have to provide a number or not set a value. null is not permitted , in this way the check if the value is set is easier

**Comment by apalumbo** [comment]
I am not sure why the CI is failing as with `lt' the tests are ok.

**Comment by apalumbo** [comment]
@rhboyd The default value for maxConcurrency (if not provided is 0 from the step functions)

> You can configure an upper bound on how many concurrent sub-workflows Map executes by adding the MaxConcurrency field. The default value is 0, which places no limit on parallelism and iterations are invoked as concurrently as possible. A MaxConcurrency value of 1 has the effect to invoke the Iterator one element at a time, in the order of their appearance in the input state, and will not start an iteration until the previous iteration has completed execution.

https://aws.amazon.com/blogs/aws/new-step-functions-support-for-dynamic-parallelism/

**Comment by rhboyd** [comment]
Awesome. Thanks for clarifying it. I was trying to read the PR on mobile and I confused myself. +1

**Comment by rhboyd** [comment]
Also, tagging @rix0rrr since he shepherded the original release into production.

**Comment by apalumbo** [comment]
@rix0rrr

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts [hunk 1]
```diff
@@ -150,10 +150,30 @@ export class Map extends State implements INextable {
...
+            const isNotPositiveInteger = maxConcurrency < 0 || maxConcurrency > Number.MAX_SAFE_INTEGER;
+
+            if (isFloat || isNotPositiveInteger) {
+                errors.push('maxConcurrency has to be a positive integer');
+            }
+        };
+
+        const errors: string[] = [];
+
+        if (this.iteration === undefined) {
```
Comment (apalumbo) [inline packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: now Map has a method named iterator (like the property of the state in the doc) and the property on the parent State object is named iteration. I think that the method or the property should be renamed

**Comment by yyolk** [comment]
I tried sync'ing a graph using this branch (only updating node-aws-cdk with `npm link`), I ended up getting an error about Iterator not being filled, copying this branch's test I also tried something like the following:
```py
m = sfn.Map(
    self,
    logical_name,
    input_path=input_path,
    output_path=output_path,
    max_concurrency=10,
    **kwargs,
)
m.iterator(sfn.Pass(self, "Pass in Map"))
```
This error'd for the same reason. I figure this is likely because I didn't also upgrade the python packages I import (cdk is only at runtime). 

I'm going through that now, running the `pack.sh` script after installing maven, python-site-packages, and dotnet. I'm also about to rap up for the day, so I'm hoping this gets merged and published before then ;)

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts [hunk 2]
```diff
@@ -150,10 +150,30 @@ export class Map extends State implements INextable {
      * Validate this state
      */
     protected validate(): string[] {
-        if (!!this.iterator) {
-            return ['Map state must have a non-empty iterator'];
+        const validateMaxConcurrency = () => {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: Not sure why this needs to be an inner function? I'd rather you extract the logic to test that a number is a positive integer to a function `isPositiveInteger()` and call that inline in this function.

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: There's something to be said for that. `iteration` is `protected` though, so will be invisible to regular consumers. What situation are you concerned about?

Comment (rix0rrr) [reply packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: Oh I see, this bug. Yes, that's unfortunate, and thanks for the fix. Given the test, I think there's no need to rename now.

File: packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts (hunk 2 above)
Comment (apalumbo) [reply packages/@aws-cdk/aws-stepfunctions/lib/states/map.ts]: Ok i will move the function outside Map and export it for the test, I will add tests for the function (the samecases  I added) but I would like to leave the tests on synth to take care about regressions

**Comment by apalumbo** [comment]
@rix0rrr It should be ok now

**Comment by apalumbo** [comment]
@rix0rrr Do you think that, is the review it's ok, there is any change that this will go in the next release? We are waiting for this feature to complete a feature in one of our projects and it would be great

**Comment by sarveshbhagat** [comment]
> I tried sync'ing a graph using this branch (only updating node-aws-cdk with `npm link`), I ended up getting an error about Iterator not being filled, copying this branch's test I also tried something like the following:
> 
> ```python
> m = sfn.Map(
>     self,
>     logical_name,
>     input_path=input_path,
>     output_path=output_path,
>     max_concurrency=10,
>     **kwargs,
> )
> m.iterator(sfn.Pass(self, "Pass in Map"))
> ```
> 
> This error'd for the same reason. I figure this is likely because I didn't also upgrade the python packages I import (cdk is only at runtime).
> 
> I'm going through that now, running the `pack.sh` script after installing maven, python-site-packages, and dotnet. I'm also about to rap up for the day, so I'm hoping this gets merged and published before then ;)

Is this working for you?

**Comment by Stvad** [comment]
Should `iterator` be added to `MapProps` ?

**Pull request merged**